### ISSUE TITLE
Throw a custom exception if Guzzle has an exception

### ISSUE THREAD

**Issue by callum-p** [configcat/php-sdk#11]
We often see these exceptions:

![image](https://user-images.githubusercontent.com/7827942/104859928-7568df00-598d-11eb-85a1-f8c0b9291886.png)

I was working on ignoring them today by blocking logging for Guzzle's `ConnectException::class` but then I wouldn't see logs if some of our inter-service communication timed out.

It would be good if Configcat could throw it's own ConnectException if Guzzle encounters an error, then I can ignore timeouts for Configcat and have less log spam.

**Comment by z4kn4fein** [comment]
Hey, thanks for reporting this issue!
Is the 1000 milliseconds timeout set by you? Could you share your client configuration?
I would focus on the timeout issue itself, before touching the logging.
Thanks!

**Comment by callum-p** [comment]
> Hey, thanks for reporting this issue!
> Is the 1000 milliseconds timeout set by you? Could you share your client configuration?
> I would focus on the timeout issue itself, before touching the logging.
> Thanks!

Hi @z4kn4fein, thanks for following up. Yes we set the 1000ms timeout ourselves as we have a queue worker processing a lot of jobs and we don't want it to be delayed with a large timeout.  We're investigating disabling feature flags for the queue worker but ideally we'd leave it on, just have the ability to tweak logging a bit.

Just for context, it's a couple of timeouts every few thousand requests, which I'd expect for any service, especially when I've set the timeout so low.

**Comment by z4kn4fein** [comment]
Ah, I see, how do you filter out Guzzle's `ConnectException` now exactly? 
I'm just thinking about what possible options we have here. What do you think about adding something to the $context array of the log entry that indicates it came from the `ConfigCatClient`?

**Comment by callum-p** [comment]
> Ah, I see, how do you filter out Guzzle's `ConnectException` now exactly?
> I'm just thinking about what possible options we have here. What do you think about adding something to the $context array of the log entry that indicates it came from the `ConfigCatClient`?

We provide our logging instance go ConfigCat like this:
```php
$options = [
                        'cache' => new LaravelCache($cache),
                        'cache-refresh-interval' => config('services.configcat.cache-refresh'),
                        'logger' => \Log::getMonolog(),
];
```

Then we have an exception handler where we can filter out unwanted logs or downgrade their severity, e.g.:

```php
/**
     * Exceptions that should only generate warning level logs instead of error level logs
     *
     * @var array
     */
    protected $warningExceptions = [
        TenantDoesNotExistException::class,
        TokenMismatchException::class,
        SSOUserNotFound::class,
    ];
...
    if (in_array(get_class($e), $this->warningExceptions)) {
            $this->logException($e, $header, 'warning');

            return;
        }
```

In order to downgrade Guzzle's ConnectException to warning or to not log them at all, I would add ConnectException::class to $warningExceptions.

**Comment by z4kn4fein** [comment]
Much clearer now, thanks! 
I'm trying to avoid wrapping the `ConnectException`, because it doesn't make much sense to me as that pattern would say we should wrap every exception with some sort of `ConfigCatClientException` to originate them from the `ConfigCatClient`. But ignoring that would mean all exceptions would be ignored. 
What do you think about a separate configuration option similar to your `$warningExceptions` like `'exceptions-to-ignore' => [ConnectException::class]`?
It gives full control over the 'to be ignored' list applied only to the ConfigCat SDK. 
Would this work for you?

**Comment by callum-p** [comment]
> Much clearer now, thanks!
> I'm trying to avoid wrapping the `ConnectException`, because it doesn't make much sense to me as that pattern would say we should wrap every exception with some sort of `ConfigCatClientException` to originate them from the `ConfigCatClient`. But ignoring that would mean all exceptions would be ignored.
> What do you think about a separate configuration option similar to your `$warningExceptions` like `'exceptions-to-ignore' => [ConnectException::class]`?
> It gives full control over the 'to be ignored' list applied only to the ConfigCat SDK.
> Would this work for you?

Yeah that would be great!

**Comment by z4kn4fein** [comment]
Released a new version [5.2.0](https://github.com/configcat/php-sdk/releases/tag/v5.2.0) with this configuration option. 
Could you please try it and let me know if there's something wrong?
Thanks!

**Comment by callum-p** [comment]
Works well @z4kn4fein , thanks a lot.
